### ISSUE TITLE
Large directory listings on `kopia mount` are very slow (1 entry per second)

### ISSUE THREAD

**Issue by nh2** [kopia/kopia#1135]
(I'm currently benchmarkiing various kopia operations.)

Directory listing on `kopia mount` is slow, e.g. when using `find`: `getdents64()` syscall on a large directory takes 20 seconds per 18 entries.

`strace -T` output lines look like:

```c
getdents64(/kopiamount/.../largedir, /* 18 entries */, 32768) = 1584 <20.273433>
```

Both the original source data to be backed up, and the kopia repository, are on localhost CephFS (kopia `--path` repository). On the source data, `find` runs on `255010` dirents in 8 seconds -- **35000x faster.**

For reference, `cat`ing a large file from `kopia mount` runs at 50 MB/s. At least in theory, large directories are also 

The snapshot that includes this directory is partial, and the `find` is under a `... (checkpoint) ...` in the `kopia mount` -- not sure if that matters. Using kopia v0.8.4 on Linux.

Is there something that could speed up directory listings on `kopia mount`?

Thanks!

**Comment by nh2** [comment]
I suspect that directory listings on `kopia mount` take quadratic time: _O(n²)_, where _n_ is the number of directories.

On a directory that has 17x less entries, the `getdents()` calls return 550x faster.

**Comment by jkowalski** [comment]
Can you try the latest build at head? There was a misplaced locks in metadata cache that might have caused it and it was fixed recently. Definitely quadratic cost is unexpected here.

**Comment by nh2** [comment]
> There was a misplaced locks in metadata cache that might have caused it and it was fixed recently

@jkowalski Will try -- which commit are you referrinig to?

**Comment by jkowalski** [comment]
https://github.com/kopia/kopia/pull/1120

**Comment by jkowalski** [comment]
You can grab latest test builds from https://github.com/kopia/kopia-test-builds/releases

**Comment by nh2** [comment]
> #1120

@jkowalski I have tested it now, unfortunately the timing is unchanged.

**Comment by jkowalski** [comment]
Can you see log entries that have STORAGE in them? What are the latencies?

**Comment by pawitp** [comment]
I took a look and found the cause. It's reproducible even on the local file system with 300k small files. However, the solution is too complex for me to tackle without guidance.

 - The FUSE command being used here is READDIRPLUS
 - When go-fuse does a READDIRPLUS, it calls `Lookup` for every file ([link](https://github.com/hanwen/go-fuse/blob/v2.1.0/fs/bridge.go#L1015))
 - Okay, we do a Lookup ([link](https://github.com/kopia/kopia/blob/v0.9.0-rc3/internal/fusemount/fusefs.go#L109)). There's two operations here, `ReadDir` and `FindByName`.
   - First I thought it's `FindByName` doing a linear search, but actually it's a binary search
   - Then I looked at `ReadDir`. We have cache so everything should be kosher, right? Unfortunately not. When reading from cache, it wraps all entry with the cache context ([link](https://github.com/kopia/kopia/blob/v0.9.0-rc3/fs/cachefs/cachefs.go#L41-L44)) which is a linear operation (resulting in `O(n^2)`!) If we move the wrapping to before we insert it into the cache, everything will be a lot faster. It does involve some refactoring between `cache.go` and `cachefs.go` though so I'm not sure what's the best way to approach it.
 - Finally, the default cache size is 100k, so you will have to run with `--max-cached-entries=1000000` if your directory is bigger than 100k otherwise the directory is no longer cached and everything grinds to a halt.

While the best way is to implement batch readdirplus so go-fuse don't call Lookup for every file, I don't see any easy way to do it without changing go-fuse itself. The second best thing to do is to fix our caching layer. With fixing the cache layer and increasing the cache size, a `ls -l` of 300k file is down to about 15 seconds.
